### PR TITLE
Roll CanvasKit to 0.16.1

### PR DESCRIPTION
**Pull request by hterkelsen**
- Fix SkColor, it migrated to a Float32List
- Fix deprecated shader calls

Fixes https://github.com/flutter/flutter/issues/58014
Fixes https://github.com/flutter/flutter/issues/56938

### CODE REVIEW

File: lib/web_ui/lib/src/engine/compositor/canvas.dart [hunk 1]
```diff
@@ -14,7 +14,7 @@ class SkCanvas {
   int get saveCount => skCanvas.callMethod('getSaveCount');
 
   void clear(ui.Color color) {
-    skCanvas.callMethod('clear', <int>[color.value]);
+    skCanvas.callMethod('clear', <js.JsObject>[makeSkColor(color)]);
```
Comment (kjlubick) [inline lib/web_ui/lib/src/engine/compositor/canvas.dart:17]: Driveby comment here from a Dart noob: Is there not a dart type for float32arrays? http://commondatastorage.googleapis.com/dartlang-api-docs/10315/dart_html/Float32Array.html

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart:17]: There is. Does `CanvasKit.Color` simply construct a `Float32Array`?

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart:17]: Looks like it is: https://github.com/google/skia/blob/41fe8bc3500ea15f1d14362ca4aeef21aecb2314/modules/canvaskit/helper.js#L26

We can do this faster in plain Dart code. However, I'm still uneasy about having to convert Flutter colors, which are plain `int` values, to `Float32Array`s :/

Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart]: I changed this to create Float32Arrays.

**Comment by hterkelsen** [comment]
PTAL, I updated the PR to use a single, Malloc'ed Float32Array where possible.

**Comment by hterkelsen** [comment]
I fixed an issue where we weren't using the shared color array for `SkPaint.setColor`. PTAL @yjbanov

File: lib/web_ui/lib/src/engine/compositor/canvas.dart [hunk 2]
```diff
@@ -97,8 +97,8 @@ class SkCanvas {
   }
 
   void drawColor(ui.Color color, ui.BlendMode blendMode) {
-    skCanvas.callMethod('drawColor', <dynamic>[
-      color.value,
+    skCanvas.callMethod('_drawColor', <dynamic>[
```
Comment (kjlubick) [inline lib/web_ui/lib/src/engine/compositor/canvas.dart:100]: This is a temporary work around because drawColor would free the Malloc'd float array, right? I think this is a strong signal that CanvasKit should not do that; I can make such changes.

Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart:100]: Yes, that's right. I added some comments clarifying why we're using the private methods here.

I suggest changing the behavior of the methods that automatically free the arrays to check if the arrays were already malloc'ed and not free them in that case.

**Comment by hterkelsen** [comment]
@yjbanov PTAL

File: lib/web_ui/lib/src/engine/compositor/util.dart [hunk 3]
```diff
@@ -152,7 +171,8 @@ js.JsArray<double> makeSkPoint(ui.Offset point) {
 Float32List encodePointList(List<ui.Offset> points) {
   assert(points != null);
   final int pointCount = points.length;
-  final Float32List result = canvasKit.callMethod('Malloc', <dynamic>[js.context['Float32Array'], pointCount * 2]);
+  final Float32List result = canvasKit.callMethod(
+      'Malloc', <dynamic>[js.context['Float32Array'], pointCount * 2]);
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/util.dart:175]: Not in 0.15.0, because everywhere the resulting list is passed (SkCanvas.drawPoints, SkCanvas.drawAtlas, SkPath.addPoly) it is automatically freed by CanvasKit. However, in the next version it will switch to not free-ing Malloc'ed arrays. I will add a TODO.

**Comment by hterkelsen** [comment]
After talking with the CanvasKit team about the API, they've made the Color APIs more flexible in 0.16, which will be released shortly. I think it would be best to change this PR to roll directly from 0.14 to 0.16 and skip 0.15 altogether.

File: lib/web_ui/lib/src/engine/compositor/canvas.dart [hunk 4]
```diff
@@ -99,10 +98,9 @@ class SkCanvas {
   }
 
   void drawColor(ui.Color color, ui.BlendMode blendMode) {
-    // Call the private '_drawColor' method rather than 'drawColor' because
-    // 'drawColor' will free the color array.
-    skCanvas.callMethod('_drawColor', <dynamic>[
-      makeSkColor(color).offsetInBytes,
+    setSharedSkColor1(color);
+    skCanvas.callMethod('drawColor', <dynamic>[
```
Comment (kjlubick) [inline lib/web_ui/lib/src/engine/compositor/canvas.dart:102]: drawColorInt should work here

Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart]: Done.

File: lib/web_ui/lib/src/engine/compositor/canvas.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart]: Changed to `drawColorInt`

File: lib/web_ui/lib/src/engine/compositor/util.dart [hunk 5]
```diff
@@ -96,6 +96,25 @@ class SkiaObjects {
   }
 }
 
+Float32List _mallocColorArray() {
+  return canvasKit
+      .callMethod('Malloc', <dynamic>[js.context['Float32Array'], 4]);
+}
+
+Float32List _colorArray;
+Float32List makeSkColor(ui.Color color, {bool useFreshArray = false}) {
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/util.dart]: Done.

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 6]
```diff
@@ -216,7 +218,8 @@ class SkTextStyle implements ui.TextStyle {
     }
 
     if (foreground != null) {
-      style['foreground'] = foreground.skiaObject;
+      style['foregroundColor'] =
+          makeSkColor(foreground.color, useFreshArray: true);
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/text.dart]: Done.

File: lib/web_ui/lib/src/engine/compositor/util.dart [hunk 7]
```diff
@@ -319,9 +339,9 @@ void drawSkShadow(
   ui.Color inAmbient = color.withAlpha((color.alpha * ambientAlpha).round());
   ui.Color inSpot = color.withAlpha((color.alpha * spotAlpha).round());
 
-  final js.JsObject inTonalColors = js.JsObject.jsify(<String, int>{
-    'ambient': inAmbient.value,
-    'spot': inSpot.value,
+  final js.JsObject inTonalColors = js.JsObject.jsify(<String, Float32List>{
+    'ambient': makeSkColor(inAmbient, useFreshArray: true),
+    'spot': makeSkColor(inSpot, useFreshArray: true),
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/util.dart]: There is a bug preventing us from using Malloc'ed arrays here.

**Comment by hterkelsen** [comment]
I resolved all of the comments and updated to `0.16.1`. I tested this against gallery and the benchmarks, PTAL

File: lib/web_ui/lib/src/engine/compositor/canvas.dart [hunk 8]
```diff
@@ -185,7 +186,8 @@ class SkCanvas {
     skCanvas.callMethod('drawPicture', <js.JsObject>[skPicture.skPicture]);
   }
 
-  void drawPoints(SkPaint paint, ui.PointMode pointMode, Float32List points) {
+  void drawPoints(SkPaint paint, ui.PointMode pointMode,
+      js.JsArray<js.JsArray<double>> points) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/canvas.dart:190]: Can you please file an issue and leave a TODO to turn this into a single flat typed array?

File: lib/web_ui/lib/src/engine/compositor/path.dart [hunk 9]
```diff
@@ -94,7 +94,7 @@ class SkPath implements ui.Path {
   @override
   void addPolygon(List<ui.Offset> points, bool close) {
     assert(points != null);
-    final Float32List encodedPoints = encodePointList(points);
+    final List<List<double>> encodedPoints = encodePointList(points);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/path.dart:97]: Issue and a TODO?

File: lib/web_ui/lib/src/engine/compositor/util.dart [hunk 10]
```diff
@@ -96,6 +96,70 @@ class SkiaObjects {
   }
 }
 
+/// Converts a list of [ui.Color] into the 2d array expected by CanvasKit.
+js.JsArray<Float32List> makeColorList(List<ui.Color> colors) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/util.dart:100]: Issue and a TODO to make this just `Float32List`?

File: lib/web_ui/lib/src/engine/compositor/canvas.dart (hunk 8 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas.dart:191]: Done.

File: lib/web_ui/lib/src/engine/compositor/path.dart (hunk 9 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/path.dart:98]: Done.

File: lib/web_ui/lib/src/engine/compositor/util.dart (hunk 10 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/util.dart:100]: Done.

**Comment by hterkelsen** [comment]
Landing despite unrelated engine failures

**Comment by luigi-rosso** [comment]
Hi @hterkelsen, I think these changes caused an issue with calling Vertices.raw without colors. _encodeRawColorList is trying to read length from a null color buffer when calling _init. The color buffer used to be optional (we've never used it before). You'll see that any projects using flare_flutter with images will fail on current master. Passing a color buffer temporarily fixes the issue.
